### ISSUE TITLE
sqrt,should convert int to float

### ISSUE THREAD

**Issue by icodingc** [bgshih/tf_resnet_cifar#2]
`stddev=math.sqrt(2/(k*k*n_in))`
this line ,you should use 2.0 instead of 2

**Comment by bgshih** [comment]
@icodingc With `from __future__ import divison`, they are the same.

**Comment by xiexiaona** [comment]
@DeadAt0m Excuse me. Regarding reproducibility of Resnet-20, I think my model architecture is correct. Because my Resnet-20 achieves 91.97% accuracy in python SSL shift operation, which is matching to the original paper. However, same model I can only achieve 89.11% accuracy in your C++ SSL shift operation.
This is the code of my python SSL shift operation. Actually, it is very slow. I write it only to verify my model architecture.

import torch
import torch.nn as nn
class Shift_py(nn.Module):
def init(self, in_channels, init_shift=3, pad_mode='zero', pad_constant=0, active_flag=False):
super(Shift_py, self).init()
self.init_shift = init_shift
self.pad_mode = pad_mode
self.pad_constant = pad_constant
self.active_flag = active_flag
if self.pad_mode == 'zero':
self.padding = nn.ZeroPad2d(self.init_shift)
elif self.pad_mode == 'replication':
self.padding = nn.ZeroPad2d(self.init_shift)
elif self.pad_mode == 'reflection':
self.padding = nn.Reflection(self.init_shift)
elif self.pad_mode == 'constant':
self.padding = nn.ConstantPad2d(self.init_shift, self.pad_constant)
self.weight = torch.empty(in_channels, 2)
nn.init.uniform_(self.weight, -self.init_shift, self.init_shift)

def forward(self, fm):
"""
weight: [C, 2]
fm: feature map [B, C, H, W]
pad_size:
pad_mode:'zero', 'replication', 'ReflectionPad2d', 'constant'
"""
self.weight = torch.clamp(self.weight, -self.init_shift, self.init_shift)
b, c, h, w = fm.size()
shift_fm = torch.zeros_like(fm)
fm = self.padding(fm)
if not self.active_flag:
for i in range(c):
dx, dy = self.weight[i]
x1 = int(self.init_shift + torch.round(dx))
y1 = int(self.init_shift + torch.round(dy))
shift_fm[:, i, :, :] = fm[:, i, y1:y1+h, x1:x1+w]

else:
for i in range(c):
x, y = self.weight[i]
x1 = int(self.init_shift + torch.floor(x))
x2 = x1 + 1
dx = self.init_shift + x - x1
y1 = int(self.init_shift + torch.floor(y))
y2 = y1 + 1
dy = self.init_shift + y - y1
shift_fm[:, i, :, :] = (1-dx)*(1-dy)fm[:, i, y1:y1+h, x1:x1+w] +
(1-dx)(dy)fm[:, i, y2:y2+h, x1:x1+w] +
(dx)(1-dy)fm[:, i, y1:y1+h, x2:x2+w] +
(dx)(dy)*fm[:, i, y2:y2+h, x2:x2+w]
return shift_fm

**Comment by xiexiaona** [comment]
@DeadAt0m Excuse me. Regarding reproducibility of Resnet-20, I think my model architecture is correct. Because my Resnet-20 achieves 91.97% accuracy in python SSL shift operation, which is matching to the original paper. However, same model I can only achieve 89.11% accuracy in your C++ SSL shift operation.
This is the code of my python SSL shift operation. Actually, it is very slow. I write it only to verify my model architecture.
`
import torch
import torch.nn as nn
class Shift_py(nn.Module):
def init(self, in_channels, init_shift=3, pad_mode='zero', pad_constant=0, active_flag=False):
super(Shift_py, self).init()
self.init_shift = init_shift
self.pad_mode = pad_mode
self.pad_constant = pad_constant
self.active_flag = active_flag
if self.pad_mode == 'zero':
self.padding = nn.ZeroPad2d(self.init_shift)
elif self.pad_mode == 'replication':
self.padding = nn.ZeroPad2d(self.init_shift)
elif self.pad_mode == 'reflection':
self.padding = nn.Reflection(self.init_shift)
elif self.pad_mode == 'constant':
self.padding = nn.ConstantPad2d(self.init_shift, self.pad_constant)
self.weight = torch.empty(in_channels, 2)
nn.init.uniform_(self.weight, -self.init_shift, self.init_shift)

def forward(self, fm):
"""
weight: [C, 2]
fm: feature map [B, C, H, W]
pad_size:
pad_mode:'zero', 'replication', 'ReflectionPad2d', 'constant'
"""
self.weight = torch.clamp(self.weight, -self.init_shift, self.init_shift)
b, c, h, w = fm.size()
shift_fm = torch.zeros_like(fm)
fm = self.padding(fm)
if not self.active_flag:
for i in range(c):
dx, dy = self.weight[i]
x1 = int(self.init_shift + torch.round(dx))
y1 = int(self.init_shift + torch.round(dy))
shift_fm[:, i, :, :] = fm[:, i, y1:y1+h, x1:x1+w]

else:
for i in range(c):
x, y = self.weight[i]
x1 = int(self.init_shift + torch.floor(x))
x2 = x1 + 1
dx = self.init_shift + x - x1
y1 = int(self.init_shift + torch.floor(y))
y2 = y1 + 1
dy = self.init_shift + y - y1
shift_fm[:, i, :, :] = (1-dx)*(1-dy)fm[:, i, y1:y1+h, x1:x1+w] +
(1-dx)(dy)fm[:, i, y2:y2+h, x1:x1+w] +
(dx)(1-dy)fm[:, i, y1:y1+h, x2:x2+w] +
(dx)(dy)*fm[:, i, y2:y2+h, x2:x2+w]
return shift_fm
`

**Comment by xiexiaona** [comment]
@DeadAt0m Excuse me. Regarding reproducibility of Resnet-20, I think my model architecture is correct. Because my Resnet-20 achieves 91.97% accuracy in python SSL shift operation, which is matching to the original paper. However, same model I can only achieve 89.11% accuracy in your C++ SSL shift operation.
This is the code of my python SSL shift operation. Actually, it is very slow. I write it only to verify my model architecture.

import torch
import torch.nn as nn
class Shift_py(nn.Module):
def init(self, in_channels, init_shift=3, pad_mode='zero', pad_constant=0, active_flag=False):
super(Shift_py, self).init()
self.init_shift = init_shift
self.pad_mode = pad_mode
self.pad_constant = pad_constant
self.active_flag = active_flag
if self.pad_mode == 'zero':
self.padding = nn.ZeroPad2d(self.init_shift)
elif self.pad_mode == 'replication':
self.padding = nn.ZeroPad2d(self.init_shift)
elif self.pad_mode == 'reflection':
self.padding = nn.Reflection(self.init_shift)
elif self.pad_mode == 'constant':
self.padding = nn.ConstantPad2d(self.init_shift, self.pad_constant)
self.weight = torch.empty(in_channels, 2)
nn.init.uniform_(self.weight, -self.init_shift, self.init_shift)

def forward(self, fm):
"""
weight: [C, 2]
fm: feature map [B, C, H, W]
pad_size:
pad_mode:'zero', 'replication', 'ReflectionPad2d', 'constant'
"""
self.weight = torch.clamp(self.weight, -self.init_shift, self.init_shift)
b, c, h, w = fm.size()
shift_fm = torch.zeros_like(fm)
fm = self.padding(fm)
if not self.active_flag:
for i in range(c):
dx, dy = self.weight[i]
x1 = int(self.init_shift + torch.round(dx))
y1 = int(self.init_shift + torch.round(dy))
shift_fm[:, i, :, :] = fm[:, i, y1:y1+h, x1:x1+w]

else:
for i in range(c):
x, y = self.weight[i]
x1 = int(self.init_shift + torch.floor(x))
x2 = x1 + 1
dx = self.init_shift + x - x1
y1 = int(self.init_shift + torch.floor(y))
y2 = y1 + 1
dy = self.init_shift + y - y1
shift_fm[:, i, :, :] = (1-dx)*(1-dy)fm[:, i, y1:y1+h, x1:x1+w] +
(1-dx)(dy)fm[:, i, y2:y2+h, x1:x1+w] +
(dx)(1-dy)fm[:, i, y1:y1+h, x2:x2+w] +
(dx)(dy)*fm[:, i, y2:y2+h, x2:x2+w]
return shift_fm

**Comment by xiexiaona** [comment]
@DeadAt0m Excuse me. Regarding reproducibility of Resnet-20, I think my model architecture is correct. Because my Resnet-20 achieves 91.97% accuracy in python SSL shift operation, which is matching to the original paper. However, same model I can only achieve 89.11% accuracy in your C++ SSL shift operation.
This is the code of my python SSL shift operation. Actually, it is very slow. I write it only to verify my model architecture.
`import torch
import torch.nn as nn
class Shift_py(nn.Module):
def init(self, in_channels, init_shift=3, pad_mode='zero', pad_constant=0, active_flag=False):
super(Shift_py, self).init()
self.init_shift = init_shift
self.pad_mode = pad_mode
self.pad_constant = pad_constant
self.active_flag = active_flag
if self.pad_mode == 'zero':
self.padding = nn.ZeroPad2d(self.init_shift)
elif self.pad_mode == 'replication':
self.padding = nn.ZeroPad2d(self.init_shift)
elif self.pad_mode == 'reflection':
self.padding = nn.Reflection(self.init_shift)
elif self.pad_mode == 'constant':
self.padding = nn.ConstantPad2d(self.init_shift, self.pad_constant)
self.weight = torch.empty(in_channels, 2)
nn.init.uniform_(self.weight, -self.init_shift, self.init_shift)
def forward(self, fm):
"""
weight: [C, 2]
fm: feature map [B, C, H, W]
pad_size:
pad_mode:'zero', 'replication', 'ReflectionPad2d', 'constant'
"""
self.weight = torch.clamp(self.weight, -self.init_shift, self.init_shift)
b, c, h, w = fm.size()
shift_fm = torch.zeros_like(fm)
fm = self.padding(fm)
if not self.active_flag:
for i in range(c):
dx, dy = self.weight[i]
x1 = int(self.init_shift + torch.round(dx))
y1 = int(self.init_shift + torch.round(dy))
shift_fm[:, i, :, :] = fm[:, i, y1:y1+h, x1:x1+w]
else:
for i in range(c):
x, y = self.weight[i]
x1 = int(self.init_shift + torch.floor(x))
x2 = x1 + 1
dx = self.init_shift + x - x1
y1 = int(self.init_shift + torch.floor(y))
y2 = y1 + 1
dy = self.init_shift + y - y1
shift_fm[:, i, :, :] = (1-dx)*(1-dy)fm[:, i, y1:y1+h, x1:x1+w] +
(1-dx)(dy)fm[:, i, y2:y2+h, x1:x1+w] +
(dx)(1-dy)fm[:, i, y1:y1+h, x2:x2+w] +
(dx)(dy)*fm[:, i, y2:y2+h, x2:x2+w]
return shift_fm
`
